File: app/src/main/java/mx/edu/upqroo/kristenandroid/data/repositories/ContactRepository.java
package mx.edu.upqroo.kristenandroid.data.repositories;

import android.app.Application;
import android.os.AsyncTask;

import java.util.List;

import androidx.lifecycle.LiveData;
import mx.edu.upqroo.kristenandroid.api.kristen.KristenApiServices;
import mx.edu.upqroo.kristenandroid.data.database.KristenRoomDatabase;
import mx.edu.upqroo.kristenandroid.data.database.daos.ContactDao;
import mx.edu.upqroo.kristenandroid.data.database.entities.Contact;

public class ContactRepository {
    private static ContactRepository mInstance;
    private ContactDao mContactDao;
    private KristenApiServices mApi;
    private Application mApp;

    private ContactRepository(Application application) {
        KristenRoomDatabase db = KristenRoomDatabase.getInstance(application);
        mContactDao = db.contactDao();
        mApi = KristenApiServices.getInstance();
        mApp = application;
    }

    public static ContactRepository getInstance(Application application) {
        if (mInstance == null) {
            mInstance = new ContactRepository(application);
        }
        return mInstance;
    }

    public LiveData<List<Contact>> getContact() {
        mApi.getContacts(mApp);
        return mContactDao.getAll();
    }

    public void updateContacts() {
        mApi.getContacts(mApp);
    }

    public List<Contact> getContactSync() {
        return mContactDao.getAllSync();
    }

    public void upsert(Contact entity) {
        long id = insert(entity);
        if (id == -1) {
            update(entity);
        }
    }

    public long insert(Contact contact) {
        return mContactDao.insert(contact);
    }

    public void deleteAll() {
        mContactDao.deleteAll();
    }

    public void update(Contact... contact){
        mContactDao.update(contact);
    }
}

File: app/src/main/java/mx/edu/upqroo/kristenandroid/data/repositories/ContactRepository.kt
package mx.edu.upqroo.kristenandroid.data.repositories

import android.app.Application
import androidx.lifecycle.LiveData
import mx.edu.upqroo.kristenandroid.api.kristen.KristenApiServices
import mx.edu.upqroo.kristenandroid.data.database.KristenRoomDatabase
import mx.edu.upqroo.kristenandroid.data.database.daos.ContactDao
import mx.edu.upqroo.kristenandroid.data.database.entities.Contact

class ContactRepository private constructor(private val mApp: Application) {
    private val mContactDao: ContactDao
    private val mApi: KristenApiServices

    val contact: LiveData<List<Contact>>
        get() {
            mApi.getContacts(mApp)
            return mContactDao.all
        }

    val contactSync: List<Contact>
        get() = mContactDao.allSync

    init {
        val db = KristenRoomDatabase.getInstance(mApp)
        mContactDao = db.contactDao()
        mApi = KristenApiServices.instance
    }

    fun updateContacts() {
        mApi.getContacts(mApp)
    }

    fun upsert(entity: Contact) {
        val id = insert(entity)
        if (id == -1L) {
            update(entity)
        }
    }

    fun insert(contact: Contact): Long {
        return mContactDao.insert(contact)
    }

    fun deleteAll() {
        mContactDao.deleteAll()
    }

    fun update(vararg contact: Contact) {
        mContactDao.update(*contact)
    }

    companion object {
        private var mInstance: ContactRepository? = null

        fun getInstance(application: Application): ContactRepository {
            if (mInstance == null) {
                mInstance = ContactRepository(application)
            }
            return mInstance as ContactRepository
        }
    }
}
